docs/lang/articles/deployment/tutorial.md

```diff
@@ -4,185 +4,227 @@ sidebar_position: 1
+We firstly write a Python script named `app.py`, which compiles the Taichi kernel as an artifact. Save the following code to your local machine and run the program, you will obtain an archived `module.tcm` in the same directory as `app.py`.
 
 ```python
+import taichi as ti
+
+ti.init(arch=ti.vulkan)
+if ti.lang.impl.current_cfg().arch != ti.vulkan:
+    raise RuntimeError("Vulkan is not available.")
+
 @ti.kernel
-def init(x: ti.types.ndarray(dtype=ti.f32, ndim=1)):
-    for i in x:
-        x[i] = 0
+def paint(n: ti.u32, t: ti.f32, pixels: ti.types.ndarray(dtype=ti.f32, ndim=2)):
+    for i, j in pixels:  # Parallelized over all pixels
+        c = ti.Vector([-0.8, ti.cos(t) * 0.2])
+        z = ti.Vector([i / n - 1, j / n - 0.5]) * 2
+        iterations = 0
+        while z.norm() < 20 and iterations < 50:
+            z = ti.Vector([z[0]**2 - z[1]**2, z[1] * z[0] * 2]) + c
+            iterations += 1
+        pixels[i, j] = 1 - iterations * 0.02
+
+mod = ti.aot.Module(ti.vulkan)
+mod.add_kernel(paint)
+mod.archive("module.tcm")
 ```
 
-After initialization, in kernel `add_base`, we want to add a floating-point number `base` to those in `x` in each frame.
+Let's dive into the code example to see what happened.
+
+We initialize Taichi specifing the backend as `ti.vulkan` at the beginning. Considering that Taichi will fall back to CPU device if the target architecture is unavailable, we check if the current backend meet our requirement.
 
 ```python
-@ti.kernel
-def add_base(x: ti.types.ndarray(ndim=1), base: ti.f32):
-    for i in range(x.shape[0]):
-        x[i] += base
+ti.init(arch=ti.vulkan)
+if ti.lang.impl.current_cfg().arch != ti.vulkan:
+    raise RuntimeError("Vulkan is not available.")
 ```
 
-You can also create an ndarray and launch the kernels in the same script to ensure they do everything you expect.
+Then, we define our Taichi kernel for computing each pixel in our program. A Taichi kernel describes two aspects of a computer program: the computation itself, and the data it operates on. Because we don't know what kind of data will be fed into the kernel before execution, we have to clearly annotate the argument types for the AOT compiler.
 
-```python
-x = ti.ndarray(ti.f32, shape=(8192))
-init(x)
+Taichi AOT module supports the following argument types: `ti.i32`, `ti.f32`, `ti.Ndarray`. Despite integers and floating-point numbers, we have a commonly-used data container called [`Ndarray`](https://docs.taichi-lang.org/api/taichi/lang/_ndarray/#taichi.lang._ndarray.Ndarray). It's similar to an [`ndarray`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html) in NumPy, or a [`Tensor`](https://pytorch.org/docs/stable/tensors.html) in PyTorch. It can be multidimensional and is laid out continuously in memory. If you have experienced the multidimensional arrays in C++, You can treat it as a nested array type like `float[6][14]`.
```

Comment:
about data types: do we want to mention textures/vectors as well?

Reply:
Let's not introducing too many concepts in this introductory documentation. They will realize they have those options scanning through the C-API docs, or the by-example tutorial when we ultimately make it an official doc.